### PR TITLE
Performance fixes for v2.0 and fix regression with out of order columns

### PR DESCRIPTION
**Pull request by pradeepnschrodinger**
## Description

This refactors some of the code in our redux logic to improve performance for v2.
The main changes are:
1. `state.columnProps` and `state.columnGroupProps` has been renamed to `columnElements` and `columnGroupElements` respectively.
2. Some usages of `_.pick`, `_.forEach`, `_.map`, `_.reduce`, etc have been hand picked and replaced with faster calls (mostly native) for performance.
3. `columnWidths` selector and `convertColumnsToElements` helper have been rewritten (keeping their structure same) to make them more efficient.
4. Immer (used by redux toolkit) [deeply freezes](https://immerjs.github.io/immer/freezing/) our redux store state. This is pretty inefficient for large data structures (eg: `elementTemplates`). Luckily, we can workaround this by manually pre-freezing them shallowly.

The refactor This also fixes a regression where FDT renders the wrong renderers (cell templates) when columns are specified in a different order than what gets rendered.
eg: 
```jsx
<Column
  columnKey="A"
  fixed={true}
/>
<Column
  columnKey="B"
  fixedRight={true}
/>
<Column
  columnKey="C"
/>
```
causes FDT to mistakenly render column C using the renderer from column B (and vice-versa).

## How Has This Been Tested?

Tested on local examples.
Particularly the "Fixed Right" example which is rendered by passing columns out of order.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.

### CODE REVIEW

File: src/FixedDataTableContainer.js [hunk 1]
```diff
@@ -150,8 +150,8 @@ class FixedDataTableContainer extends React.Component {
   static getBoundState(reduxStore) {
     const state = reduxStore.getState();
     const boundState = pick(state, [
-      'columnGroupProps',
-      'columnProps',
+      'columnGroupElements',
+      'columnElements',
```
Comment (pradeepnschrodinger) [inline src/FixedDataTableContainer.js:154]: `state.columnProps` has been replaced with `state.columnElements`.

One reason being that there's a modified version of `columnProps` returned by selectors, leading to confusion between `state.columnProps` and `someSelector().columnProps`.

File: src/helper/convertColumnElementsToData.js [hunk 2]
```diff
@@ -12,27 +12,42 @@
...
 
 import React from 'react';
-import forEach from 'lodash/forEach';
 import invariant from '../stubs/invariant';
-import map from 'lodash/map';
-import pick from 'lodash/pick';
+import { CellGroupType } from '../enums/CellGroup';
 
+// NOTE (pradeep): This can be simplified via _.pick()
+// However, _.pick is much slower than the hand written version here.
```
Comment (pradeepnschrodinger) [inline src/helper/convertColumnElementsToData.js:19]: https://www.measurethat.net/Benchmarks/Show/6813/0/lodash-pick-vs-native-destructuring-vs-manual

File: src/helper/convertColumnElementsToData.js [hunk 3]
```diff
@@ -58,47 +100,53 @@ function convertColumnElementsToData(childComponents) {
...
-    };
+
+  let columnIndex = 0;
+  let columnGroupIndex = 0;
+
+  if (useGroupHeader) {
+    for (const columnGroupReactElement of children) {
+      const cellGroupType = getCellGroupType(columnGroupReactElement);
+      const columnGroupProps = _extractProps(columnGroupReactElement);
+      columnGroupProps.index = columnGroupIndex;
```
Comment (pradeepnschrodinger) [inline src/helper/convertColumnElementsToData.js:116]: Also calculate and store the index while extracting the column/column-group.

File: src/helper/convertColumnElementsToData.js [hunk 4]
```diff
@@ -58,47 +100,53 @@ function convertColumnElementsToData(childComponents) {
...
-  // Use a default column group
-  forEach(children, (child) => {
-    columnProps.push(_extractProps(child));
-    _extractTemplates(elementTemplates, child);
-  });
   return {
-    columnGroupProps: [],
-    columnProps,
+    columnGroupElements,
+    columnElements,
```
Comment (pradeepnschrodinger) [inline src/helper/convertColumnElementsToData.js:147]: `columnGroupElements` groups column props by `cellGroupType`, where as `columnProps` is a flat ordered list of `column prop` objects.

File: src/helper/widthHelper.js [hunk 5]
```diff
@@ -12,16 +12,55 @@
 
 'use strict';
 
-import reduce from 'lodash/reduce';
+import { createSelector } from 'reselect';
```
Comment (pradeepnschrodinger) [inline src/helper/widthHelper.js:15]: `createSelector` is used here to memoize the selectors.

File: src/helper/widthHelper.js [hunk 6]
```diff
@@ -12,16 +12,55 @@
...
-}
+const emptyArray = [];
+const emptyObject = {};
 
-export function getTotalWidth(columns) {
-  return reduce(columns, (accum, column) => accum + column.width, 0);
-}
+const cacheOptions = {
+  memoizeOptions: {
+    maxSize: 12, // 3 cell group types * 4 template types
```
Comment (pradeepnschrodinger) [inline src/helper/widthHelper.js:22]: This ensures we can call the selectors for each cell group type for each template type without having to recompute the widths.

File: src/selectors/columnWidths.js [hunk 7]
```diff
@@ -164,55 +175,18 @@ function flexWidths(columnGroupProps, columnProps, viewportWidth) {
  *   scrollableColumns: !Array.<columnDefinition>
  * }}
  */
-function groupElements(elements) {
```
Comment (pradeepnschrodinger) [inline src/selectors/columnWidths.js:167]: Besides grouping, `groupElements` also assigned `index` and `offset`.
The extra work is no longer needed.

File: src/selectors/columnWidths.js [hunk 8]
```diff
@@ -105,54 +111,59 @@ function columnWidths(
...
-    let remainingFlexWidth = Math.max(viewportWidth - columnsWidth, 0);
-
-    // calculate and set width for each column
-    newColumnProps = map(columnProps, (column) => {
-      const { flexGrow } = column;
-
-      // if no flexGrow is specified, column defaults to original width
-      if (!flexGrow) {
-        return column;
+function flexWidths(columnGroupElements, columnElements, viewportWidth) {
```
Comment (pradeepnschrodinger) [inline src/selectors/columnWidths.js:114]: The main change here is that the `flexWidths` also calculates the offsets of the column based on the new flex width.

This was previously done by `groupElements()` below.

File: test/helper/convertColumnElementsToData-test.js [hunk 9]
```diff
@@ -6,6 +6,26 @@ import forEach from 'lodash/forEach';
 
 import convertColumnElementsToData from '../../src/helper/convertColumnElementsToData';
 
+function column(overrides) {
```
Comment (pradeepnschrodinger) [inline test/helper/convertColumnElementsToData-test.js:9]: helper to create a column element

File: yarn.lock [hunk 10]
```diff
@@ -7818,7 +7818,12 @@ requires-port@^1.0.0:
   resolved "https://registry.yarnpkg.com/requires-port/-/requires-port-1.0.0.tgz#925d2601d39ac485e091cf0da5c6e694dc3dcaff"
   integrity sha1-kl0mAdOaxIXgkc8NpcbmlNw9yv8=
 
-reselect@^4.0.0, reselect@^4.1.5:
+reselect@^4.0.0:
+  version "4.1.8"
```
Comment (pradeepnschrodinger) [inline yarn.lock:7822]: Bumped to latest version which supports cache customization for the selectors.

File: src/reducers/index.js [hunk 11]
```diff
@@ -152,7 +152,7 @@ const slice = createSlice({
     },
     propChange(state, action) {
       const { newProps, oldProps } = action.payload;
-      const oldState = clone(state);
+      const oldState = clone(original(state));
```
Comment (karry08) [inline src/reducers/index.js:155]: what does original do

File: src/reducers/index.js [hunk 12]
```diff
@@ -244,16 +244,32 @@ function initializeRowHeightsAndOffsets(state) {
...
+  //
+  // From Immer's docs:
+  //     Immer freezes everything recursively. For large data objects
+  //     that won't be changed in the future this might be over-kill,
+  //     in that case it can be more efficient to shallowly
+  //     pre-freeze data using the freeze utility.
+  Object.freeze(state.columnElements);
+  Object.freeze(state.columnGroupElements);
+  Object.freeze(state.elementTemplates);
+
```
Comment (karry08) [inline src/reducers/index.js:272]: nice

File: src/reducers/index.js (hunk 11 above)
Comment (pradeepnschrodinger) [reply src/reducers/index.js:155]: `state` is proxy object from immer that points to the actual redux state. 
Iterating over this proxy object is slightly slower than iterating over the original state.

`original(proxyObject)` returns back the original object pointed to by the proxy.
(See [original](https://redux-toolkit.js.org/api/other-exports#original) docs)

File: src/helper/convertColumnElementsToData.js [hunk 13]
```diff
@@ -58,48 +100,56 @@ function convertColumnElementsToData(childComponents) {
...
 
-  const elementTemplates = {
-    cell: [],
-    footer: [],
-    groupHeader: [],
-    header: [],
-  };
+  const columnElements = getElementsContainer();
+  const columnGroupElements = getElementsContainer();
+  const elementTemplates = getElementTemplates();
```
Comment (karry08) [inline src/helper/convertColumnElementsToData.js:105]: just a thought shouldnt the function name be getEmptyElementContainer/ getDefaultElementContainer

or what you have done is fine?

File: src/helper/widthHelper.js [hunk 14]
```diff
@@ -12,16 +12,55 @@
...
+  (columns) => columns || emptyArray,
+  (columns) => {
+    return columns.reduce((accum, column) => accum + column.props.width, 0);
+  },
+  cacheOptions
+);
 
-export function getTotalFlexGrow(columns) {
-  return reduce(columns, (accum, column) => accum + (column.flexGrow || 0), 0);
+export const getTotalWidth = createSelector(
```
Comment (karry08) [inline src/helper/widthHelper.js:34]: isnt getTotalWidth name a little confusing

**Comment by karry08** [comment]
LGTM

File: src/reducers/index.js [hunk 15]
```diff
@@ -244,16 +244,32 @@ function initializeRowHeightsAndOffsets(state) {
...
+  // performance bottle necks
+  //
+  // From Immer's docs:
+  //     Immer freezes everything recursively. For large data objects
+  //     that won't be changed in the future this might be over-kill,
+  //     in that case it can be more efficient to shallowly
+  //     pre-freeze data using the freeze utility.
+  Object.freeze(state.columnElements);
+  Object.freeze(state.columnGroupElements);
+  Object.freeze(state.elementTemplates);
```
Comment (AmanGupta2708) [inline src/reducers/index.js:271]: Can  `convertColumnElementsToData` itself return these objects as frozen???

**Comment by AmanGupta2708** [comment]
will it merge to column-virtualization as well ?

File: src/reducers/index.js (hunk 15 above)
Comment (pradeepnschrodinger) [reply src/reducers/index.js:271]: We could, but I like to keep `convertColumnElementsToData` as a pure helper method that shouldn't know about immer/parent reducer details.

File: src/helper/convertColumnElementsToData.js (hunk 13 above)
Comment (pradeepnschrodinger) [reply src/helper/convertColumnElementsToData.js:105]: hmm both seem fine to me, but I'll rename it to `getEmptyElementsContainer` to be explicit.

File: src/helper/convertColumnElementsToData.js [hunk 16]
```diff
@@ -58,48 +100,56 @@ function convertColumnElementsToData(childComponents) {
...
+  const useGroupHeader = !!(
+    children.length && children[0].type.__TableColumnGroup__
```
[the remark this answers is not in the document]
Comment (pradeepnschrodinger) [reply src/helper/convertColumnElementsToData.js]: done

**Comment by pradeepnschrodinger** [comment]
> will it merge to column-virtualization as well ?

@AmanGupta2708 , yep this should be upmerged manually.
I don't think it'll be an easy merge though.
There are also lots of other commits in v2 that needs upmerging. 
It makes sense to do them all, and perform another round of performance benchmarks.

**Pull request merged**